pages/doc.js

```diff
@@ -25,216 +24,146 @@ import { media } from '../src/styles'
-                isMenuOpen: false,
-                markdown: text,
-                pageNotFound: false
-              },
-              () => this.updateScroll(!isFirstPage && isPageChanged)
-            )
-          })
-        })
-        .catch(() => {
-          window.location.reload()
-        })
-    }
-  }
+        <Side isOpen={isMenuOpen}>
+          {isSearchAvaible && (
+            <SearchArea>
+              <SearchForm />
+            </SearchArea>
+          )}
 
-  updateScroll(isPageChanged) {
-    const { hash } = window.location
+          <SidebarMenu sidebar={sidebar} currentPath={path} id={SIDEBAR_MENU} />
+        </Side>
+
+        <Markdown
+          markdown={markdown}
+          githubLink={githubLink}
+          prev={prev}
+          next={next}
+          tutorials={tutorials}
+        />
+
+        <RightPanel
+          headings={headings}
+          githubLink={githubLink}
+          tutorials={tutorials}
+        />
+      </Container>
+    </Page>
+  )
+}
 
-    if (isPageChanged) {
-      this.setState({ isSmoothScrollEnabled: false }, () => {
-        this.scrollTop()
-        this.setState({ isSmoothScrollEnabled: true }, () => {
-          if (hash) this.scrollToLink(hash)
-        })
-      })
-    } else if (hash) {
-      this.scrollToLink(hash)
+Documentation.getInitialProps = async ({ asPath, req }) => {
+  const item = getItemByPath(asPath)
+
+  if (!item) {
+    return {
+      errorCode: 404
     }
   }
 
-  parseHeadings = text => {
-    const headingRegex = /\n(## \s*)(.*)/g
-    const matches = []
-    let match
-    do {
-      match = headingRegex.exec(text)
-      if (match)
-        matches.push({
-          text: match[2],
-          slug: kebabCase(match[2])
-        })
-    } while (match)
-
-    return matches
-  }
+  const host = req ? req.headers['host'] : window.location.host
+  const protocol = host.indexOf('localhost') > -1 ? 'http:' : 'https:'
 
-  scrollToLink = hash => {
-    const element = document.getElementById(hash.replace(/^#/, ''))
+  try {
+    const res = await fetch(`${protocol}//${host}${item.source}`)
 
-    if (element) {
-      const headerHeight = document.getElementById(HEADER).offsetHeight
-      const elementBoundary = element.getBoundingClientRect()
-      const rootElement = document.getElementById(ROOT_ELEMENT)
-      rootElement.scroll({ top: elementBoundary.top - headerHeight })
+    if (res.status !== 200) {
+      return {
+        errorCode: res.status
+      }
     }
-  }
 
-  scrollTop = () => {
-    const rootElement = document.getElementById(ROOT_ELEMENT)
-    if (rootElement) {
-      rootElement.scrollTop = 0
-    }
-  }
+    const text = await res.text()
 
-  toggleMenu = () => {
-    this.setState(prevState => ({
-      isMenuOpen: !prevState.isMenuOpen
-    }))
+    return {
+      item: item,
+      headings: parseHeadings(text),
+      markdown: text
+    }
+  } catch (e) {
+    return {
+      errorCode: 404
+    }
   }
```

Comment:
Hmmm. what kind of errors can `fetch` throw that would get caught here? Can't easily tell from https://github.com/matthew-andrews/isomorphic-fetch

Reply:
`fetch` is a promise. It can return either `Promise.resolve` or `Promise.reject`. Resolve is the thing that we will receive in the most cases. It have `response` object and we can check it's `status` field. But if, for example, your internet connection is lost, the browser will not resolve it, but  will instead throw error which we are catching there.

Try to set connection to offline in Network tab of Developers Tools and then try to click on the internal link, it will throw error.

Reply:
Hmmm OK but why 404? I see this is changed though... Will re-ask.